Sprint 1/step 2

Ekaterina64 opened the pull request.

Ekaterina-ux committed:
Добавлен кастомный хук useFilter
Подключен api

Ekaterina-ux committed:
Добавлено модальное окно с деталями ингредиента

Ekaterina-ux committed:
Добавлено модальное окно OrderDetails

andryxxa93 commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,81 @@
+    fetch(API_INGREDIENTS_URL)
+      .then(res => res.ok ? res.json() : Promise.reject(res))
+      .then(({data}) => setData(data))
+      .catch((e) => {
```
**Можно лучше**
хорошо, что обрабатывается ошибка запроса к фзш api, но можно сделать еще лучше - сохранять в state и показывать пользователю ошибку в интерфейсе, чтобы он понимал что происходит

andryxxa93 commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,81 @@
+
+  function getData() {
+    fetch(API_INGREDIENTS_URL)
+      .then(res => res.ok ? res.json() : Promise.reject(res))
```
**Можно лучше**

в дальнейшем добавятся ещё запросы к бэкенду и лучше вынести описание этих запросов в функции

Например:

```
function getIngredients() {
   return fetch(`${NORMA_API}/ingredients`)
    .then(проверка ответа)
}

```

и разместить их в отдельном файле  `utils/burger-api.js`, а не описывать запросы в самом компоненте

andryxxa93 replied:
**Можно лучше**

Т.к. эта проверка нужна у всех запросов, лучше вынести её в отдельную функцию

```
const checkReponse = (res) => {
  return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
};
```

`fetch(someUrl).then(checkReponse)`

andryxxa93 commented on src/components/app/app.jsx:
```diff
@@ -1,17 +1,81 @@
+            <BurgerIngredients data={data} onClick={handleClick}/>
+            <BurgerConstructor onClick={handleButtonClick}/>
+          </div>
+          <Modal
```
**Необходимо исправить**
Не нужно передавать в модалку `isOpened` пропс, достаточно на этом уровне не рендерить всю модалку если этот параметр false.

```jsx
{ isModalOrderDetailsOpened &&  (
        <Modal
            title=''
            isOpened={isModalOrderDetailsOpened}
            onClose={onModalOrderDetailsClose}
          >
            <OrderDetails/>
          </Modal>
)}
```

andryxxa93 commented on src/components/burger-constructor/filling-list/filling-list.jsx:
```diff
@@ -1,42 +1,27 @@
 	);
 };
 
+FillingList.propTypes = {
```
**Необходимо исправить**
Сейчас получается, что под `fillingList` может прийти массив любого содержания, но это не так. Необходимо полностью расписать тип ингредиента со всеми его полями (чтобы не дублировать код - сделайте это в отдельном файле) и использовать во всех подобных местах. Вам помогут в этом функции propTypes `arrayOf` и `shape`.

andryxxa93 reviewed (changes requested):
Здравствуйте! Проделана большая работа 👍 

Сильные стороны проекта:
- Есть проверка успешности запроса
- Правильно использовали useEffect
- Сделали отписку от слушателя событий

Однако есть пару моментов, которые необходимо исправить, чтобы работа была принята.

Необходимо исправить:
- Не нужно передавать в модалку проп `isOpened`, достаточно на уровне `App` не рендерить всю модалку если этот параметр false.
- Распишите тип ингредиента и используйте везде где он принимается в качестве пропсов

Более подробно описание проблемы смотрите комментарии в pull request.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

**Напомню, чтобы работа была принята нужно исправить все замечания помеченные как **необходимо исправить**.

Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12) перед отправкой на ревью. 

Удачных правок!

Ekaterina-ux committed:
Исправлено:
- рендер модальных по открытию
- использование типов полей ингредиента в качестве пропсов

Ekaterina-ux committed:
Улучшено:
- Ошибка запроса в state и интерфейсе
- Описание запроса в функции
- Проверка запроса в функции

andryxxa93 commented on src/utils/prop-types.js:
```diff
@@ -0,0 +1,16 @@
+import PropTypes from "prop-types"
+
+export const IngredientPropType = {
+	_id: PropTypes.string,
```
**Можно лучше**
У полей объекта так же нужно ставить `isRequired`

andryxxa93 reviewed (approved):
Супер, вы отлично поработали!
Поздравляю! Работа принята.
Желаю удачного продолжения обучения!
